File: Graph Data Structures/C++/DFS Traversal.cpp
#include <iostream>

using namespace std;

void push( int x, int stack[], int &ptr );
int pop( int stack[] , int &ptr );

int main()
{
    int N;

    cout << "Enter No vertices in Graph : ";
    cin >> N;

    int graph[N+1][N+1];
    int flag[N+1];
    int stack[N];
    int ptr = -1;
    int x;

    for( int i = 1; i <= N; i++ )
    {
        flag[i] = -1;
    }

    cout << endl << "Enter Values of edges in Matrix" << endl;
    for( int i = 1; i <= N; i++ )
    {
        for( int j = 1; j <= N; j++ )
        {
            cin >> graph[i][j];
        }
    }

    cout << endl << "Enter the vertex to start from : ";
    cin >> x;

    push( x, stack, ptr );
    flag[x]++;

    cout << "DFS Traversal is : ";
    while( ptr != -1 )
    {
        if( flag[x] == 0 )
        {
            cout << x << " ";
            flag[x]++;
        }

        int i = 1;
        while( i < N+1 )
        {
            if( graph[x][i] > 0 && flag[i] == -1 && i <= N )
            {
                break;
            }
            i++;
        }

        if( i == N + 1 )
        {
            pop( stack, ptr );
        }else
        {
            push( i, stack, ptr );
            flag[i]++;
        }

        x = stack[ptr];
    }


    /*cout << "The given matrix is" << endl;
    for( int i = 1; i <= N; i++ )
    {
        for( int j = 1; j <= N; j++ )
        {
            cout << graph[i][j] << " ";
        }
        cout << endl;
    }*/

    return 0;
}

void push( int x, int stack[], int &ptr )
{
    ptr++;
    stack[ptr] = x;
}

int pop( int stack[] , int &ptr )
{
    int x = stack[ptr];
    ptr--;
    return x;
}

File: Graph Data Structures/Python/DFS Traversal.py
def push (data, stack):
    global ptr
    ptr += 1
    stack.append(data)

def pop(stack):
    global ptr
    x = stack[ptr]
    ptr -= 1
    return x

N = int(input("Enter No of Vertices in Graph : "))

graph = []
stack = []
ptr = -1
flag = [0]

print("Enter value of edges in matrix")
for i in range(N+1):
    arr = []
    if i == 0:
        for j in range(N+1):
            arr.append(0)
    else:
        for j in range(N+1):
            if j == 0:
                arr.append(0)
            else:
                print("Enter value in ", i, "th row and",j,"th column : ", end="")
                arr.append(int(input()))
    graph.append(arr)

for i in range(N+1):
    flag.append(-1)

startVertex = int(input("Enter Vertex to start from : "))
push(startVertex, stack)
flag[startVertex] += 1

print("DFS Traversal is : ", end="")
while( ptr != -1 ):
    if flag[startVertex] == 0:
        print(startVertex," ", end ="")
        flag[startVertex] += 1

    i = 1
    while i < N+1:
        if i <= N and graph[startVertex][i] > 0 and flag[i] == -1:
            break
        i += 1

    if i == N+1:
        pop(stack)
    else:
        push(i, stack)
        flag[i] += 1
    startVertex = stack[ptr]
